R/ProcessCSV.R

```diff
@@ -345,6 +452,246 @@ ProcessCSV <- function(user_csv, user_action, sample_storage_type, search_type =
+          dplyr::rename(strain = name) %>%
+          dplyr::mutate(strain = toupper(strain))
+        , by = c("strain")
+      ) %>%
+      filter(!is.na(id)) %>%
+      select(all_of(colnames(user_data))) %>%
+      collect()
+
+    err <- .maybe_add_err(err, df, "Strain detected", dbmap)
+
+  },
+  warning = function(w) {
+    bError <<- TRUE
+    errmsg <<- w$message
+  },
+  error = function(e) {
+    bError <<- TRUE
+    errmsg <<- e$message
+  },
+  finally = {
+
+    if (!is.null(con)) {
+      dbDisconnect(con)
+    }
+
+    if (bError) {
+      stop(errmsg)
+    }
+
+    ## throw if bad data found
+    if (!is.null(err) && length(err) > 0) {
+      stop_validation_error("There was a problem with the content of your file.", err)
+    }
+  })
+
+  return(user_data)
+}
+
+ValidateReferenceControls <- function(database, user_data, dbmap) {
+  required_names <- requires_data <- NULL
+
+  required_names <- c("strain", "percentage", "count", "density", "batch", "manifest_name", "name", "level_I", "level_II")
+
+  if (is.null(required_names)) {
+    stop("Invalid reference detected")
+  }
+
+  requires_data <- c(requires_data, required_names)
+
+  bError <- FALSE
+  err <- errmsg <- NULL
+
+  con <- NULL
+  tryCatch({
+
+    for (nm in names(dbmap)) {
+      if (all(dbmap[[nm]] %in% colnames(user_data))) {
+        user_data = dplyr::rename(user_data, unlist(dbmap[nm]))
+      } else {
+        user_data[[nm]] = c(NA)
+      }
+    }
+
+    user_data = dplyr::mutate(user_data, row_number = row_number())
+
+    stopifnot("Required database column names not implemented" = !is.null(required_names))
+
+    matched_columns <- match(required_names, names(user_data))
+    if (any(is.na(matched_columns))) {
+      errmsg <- paste("Required database column is missing:", paste(required_names[is.na(matched_columns)], collapse = ", "))
+      stop(errmsg)
+    }
+
+    ## broad sweep check if missing any data in required fields
+    rs <- rowSums(is.na(user_data[, requires_data])) > 0
+    rn <- user_data[rs, ] %>% pull(row_number)
+
+    cs <- colSums(is.na(user_data[, requires_data])) > 0
+    cols <- colnames(user_data[, requires_data])[cs]
+
+    if (length(cols) > 0) {
+      df <- user_data[rn, c("row_number", cols)]
+      err <- .maybe_add_err(err, df, "Rows found with missing data", dbmap)
+    }
+
+    ## detect `k` or `K` and convert to numeric
+    vb <- grepl('k|K', user_data$density)
+    user_data$density[vb] <- str_replace(user_data$density[vb], "k|K", "")
+    user_data$density <- as.integer(user_data$density)
+    user_data$density[vb] <- user_data$density[vb] * 1000 # convert to numeric
+
+
+    ## check date formatting
+    # allowed_date_formats = c("%Y-%m-%d")
+    # parsed_dates <- lubridate::parse_date_time(user_data$date, allowed_date_formats, quiet = TRUE, exact = TRUE)
+    # err <- .maybe_add_err(err, user_data[is.na(parsed_dates), c("row_number", "date")], "Unrecognized date format.", dbmap)
+
+    user_data.1 = user_data %>%
```

Comment:
format for entering a strain is:

|Strain|Percentage|
|W2;3d7|60;40|

should document this.  What happens if the number of strains and percentages doesn't match?